server/src/main/java/keywhiz/service/daos/ClientDAO.java

```diff
@@ -83,18 +87,37 @@ public void deleteClient(Client client) {
     });
   }
 
-  public void sawClient(Client client) {
+  public void sawClient(Client client, @Nullable Principal principal) {
     Instant now = Instant.now();
-    Instant lastSeen = Optional.ofNullable(client.getLastSeen()).map(ls -> Instant.ofEpochSecond(ls.toEpochSecond())).orElse(null);
 
-    // only update last seen if it's been more than `lastSeenThreshold` seconds
-    // this way we can have less granularity on lastSeen and save db writes
-    if (lastSeen == null || now.isAfter(lastSeen.plus(lastSeenThreshold, SECONDS))) {
+    Instant lastSeen = Optional.ofNullable(client.getLastSeen())
+        .map(ls -> Instant.ofEpochSecond(ls.toEpochSecond()))
+        .orElse(EPOCH);
+
+    final Instant expiration;
+    if (principal instanceof CertificatePrincipal) {
+      expiration = Optional.ofNullable((CertificatePrincipal) principal)
+          .map(p -> p.getCertificateExpiration())
+          .orElse(EPOCH);
+    } else {
+      expiration = EPOCH;
+    }
+
+    // Only update last seen if it's been more than `lastSeenThreshold` seconds
+    // this way we can have less granularity on lastSeen and save DB writes
+    if (now.isAfter(lastSeen.plus(LAST_SEEN_THRESHOLD))) {
       dslContext.transaction(configuration -> {
-        Param<Long> val = DSL.val(now.getEpochSecond(), CLIENTS.LASTSEEN);
+        Param<Long> lastSeenValue = DSL.val(now.getEpochSecond(), CLIENTS.LASTSEEN);
+        Param<Long> expirationValue = DSL.val(expiration.getEpochSecond(), CLIENTS.EXPIRATION);
+
         DSL.using(configuration)
             .update(CLIENTS)
-            .set(CLIENTS.LASTSEEN, DSL.when(CLIENTS.LASTSEEN.isNull(), val).otherwise(DSL.greatest(CLIENTS.LASTSEEN, val)))
+            .set(CLIENTS.LASTSEEN,
+                when(CLIENTS.LASTSEEN.isNull(), lastSeenValue)
+                    .otherwise(greatest(CLIENTS.LASTSEEN, lastSeenValue)))
+            .set(CLIENTS.EXPIRATION,
+                when(CLIENTS.EXPIRATION.isNull(), expirationValue)
+                    .otherwise(greatest(CLIENTS.EXPIRATION, expirationValue)))
```

Comment:
This assumes no client will ever present a certificate which expires before a certificate it used previously.  It might be safer to always overwrite the recorded expiration with the just-seen expiration.

Reply:
Good point; updated.